File: tools/jev-browser/core/src/planner.ts
import type { ActionStep, FlowStep } from './types.js';
import { err } from './errors.js';
import type { PlannerConfig } from './config.js';

/**
 * 可选规划器（DESIGN §7 / P4）：`run` 模式由它把完整目标拆成受 schema 校验的 FlowStep。
 * 首个 Provider 为 openai-compatible（用户显式配置 baseUrl/model/keyEnv；无默认厂商）。
 * P4 退出门槛要求实现前核对真实服务协议；本实现标注为需 P4 联调验证。
 */

export interface PlannerInput {
  goal: string;
  successCriteria: string;
  valuesKeys: string[];
  allowedOrigins: string[];
  currentUrl?: string;
}

export interface PlannerProvider {
  plan(input: PlannerInput): Promise<FlowStep[]>;
}

const STEP_LIMIT = 30;

function systemPrompt(): string {
  return [
    '你是浏览器自动化任务规划器。把目标和验收条件拆成严格的 JSON 步骤数组。',
    '输出 JSON：{"steps": FlowStep[]}。FlowStep 的 kind 只能是 action|assert|extract|branch|forEach|goal。',
    '规则：',
    '1. action.action 只能是 navigate|click|fill|press|select|scroll|wait|screenshot。',
    '2. LocatorSpec.by 只能是 role|label|testId|text|css；优先 role+name。',
    '3. navigate/click/fill 等 action 必须带非空 expect 后置条件。',
    '4. branch 只允许一层（then 内不得再有 branch/forEach）；forEach.maxItems 不得超过 30。',
    `5. 总步骤数不得超过 ${STEP_LIMIT}；优先 extract+assert 组合而不是猜测。`,
    '6. 不要发明 values 中不存在的键；敏感值用 valuesRef 引用。',
    '7. 目标不可达或信息不足时，输出 {"steps":[]}。',
  ].join('\n');
}

function userPrompt(input: PlannerInput): string {
  return JSON.stringify({
    goal: input.goal,
    successCriteria: input.successCriteria,
    valuesKeys: input.valuesKeys,
    allowedOrigins: input.allowedOrigins,
    currentUrl: input.currentUrl,
  });
}

const ACTION_NAMES = new Set(['navigate', 'click', 'fill', 'press', 'select', 'scroll', 'wait', 'screenshot']);
const LOCATOR_BY = new Set(['role', 'label', 'testId', 'text', 'css']);
const EXPECT_KINDS = new Set(['url_contains', 'text_present', 'visible', 'hidden', 'count_gte', 'download_completed', 'var_equals']);

function validateLocator(raw: unknown, where: string): void {
  if (typeof raw !== 'object' || raw === null) throw err('PLANNER_INVALID_OUTPUT', `${where} 必须是对象`);
  const spec = raw as Record<string, unknown>;
  if (!LOCATOR_BY.has(String(spec.by))) throw err('PLANNER_INVALID_OUTPUT', `${where}.by 非法: ${String(spec.by)}`);
}

function validateExpects(raw: unknown, where: string): ActionStep['expect'] {
  if (raw === undefined) return [];
  if (!Array.isArray(raw)) throw err('PLANNER_INVALID_OUTPUT', `${where} 必须是数组`);
  return raw.map((e, i) => {
    if (typeof e !== 'object' || e === null) throw err('PLANNER_INVALID_OUTPUT', `${where}[${i}] 必须是对象`);
    const spec = e as Record<string, unknown>;
    if (!EXPECT_KINDS.has(String(spec.kind))) {
      throw err('PLANNER_INVALID_OUTPUT', `${where}[${i}].kind 非法: ${String(spec.kind)}`);
    }
    return spec as unknown as ActionStep['expect'][number];
  });
}

/** 严格校验规划输出：模型只能产出白名单步骤（DESIGN §7）。 */
export function validatePlannedSteps(raw: unknown, depth = 0): FlowStep[] {
  if (!Array.isArray(raw)) throw err('PLANNER_INVALID_OUTPUT', 'steps 必须是数组');
  if (raw.length > STEP_LIMIT) throw err('PLANNER_INVALID_OUTPUT', `步骤数 ${raw.length} 超过上限 ${STEP_LIMIT}`);
  return raw.map((s, idx): FlowStep => {
    if (typeof s !== 'object' || s === null) throw err('PLANNER_INVALID_OUTPUT', `steps[${idx}] 不是对象`);
    const step = s as Record<string, unknown>;
    const id = typeof step.id === 'string' && step.id ? step.id : `p${idx}`;
    switch (step.kind) {
      case 'action': {
        const action = String(step.action);
        if (!ACTION_NAMES.has(action)) throw err('PLANNER_INVALID_OUTPUT', `steps[${idx}].action 非法: ${action}`);
        if (step.target !== undefined) validateLocator(step.target, `steps[${idx}].target`);
        const expect = validateExpects(step.expect, `steps[${idx}].expect`);
        if (expect.length === 0) throw err('PLANNER_INVALID_OUTPUT', `steps[${idx}] action 缺少 expect 后置条件`);
        const out: ActionStep = {
          id,
          kind: 'action',
          action: action as ActionStep['action'],
          expect,
        };
        if (step.target !== undefined) out.target = step.target as ActionStep['target'];
        if (typeof step.value === 'string' || typeof step.value === 'number') out.value = step.value;
        if (typeof step.valuesRef === 'string') out.valuesRef = step.valuesRef;
        if (typeof step.key === 'string') out.key = step.key;
        return out;
      }
      case 'assert':
        return { id, kind: 'assert', expect: validateExpects(step.expect, `steps[${idx}].expect`) };
      case 'extract': {
        validateLocator(step.target, `steps[${idx}].target`);
        const fields = Array.isArray(step.fields) ? (step.fields.filter((f) => f === 'text' || f === 'count') as Array<'text' | 'count'>) : [];
        if (fields.length === 0) throw err('PLANNER_INVALID_OUTPUT', `steps[${idx}].fields 非法`);
        if (typeof step.saveAs !== 'string' || !step.saveAs) throw err('PLANNER_INVALID_OUTPUT', `steps[${idx}].saveAs 非法`);
        return { id, kind: 'extract', target: step.target as never, fields, saveAs: step.saveAs };
      }
      case 'branch': {
        if (depth > 0) throw err('PLANNER_INVALID_OUTPUT', 'branch 不允许嵌套');
        if (typeof step.variable !== 'string') throw err('PLANNER_INVALID_OUTPUT', `steps[${idx}].variable 非法`);
        const then = validatePlannedSteps(step.then, depth + 1);
        return { id, kind: 'branch', variable: step.variable, equals: step.equals as string | number, then };
      }
      case 'forEach': {
        if (depth > 0) throw err('PLANNER_INVALID_OUTPUT', 'forEach 不允许嵌套');
        const maxItems = Number(step.maxItems);
        if (!Number.isFinite(maxItems) || maxItems < 1 || maxItems > 30) {
          throw err('PLANNER_INVALID_OUTPUT', `steps[${idx}].maxItems 必须在 1..30`);
        }
        if (typeof step.itemsVar !== 'string' || typeof step.itemVar !== 'string') {
          throw err('PLANNER_INVALID_OUTPUT', `steps[${idx}] itemsVar/itemVar 非法`);
        }
        const body = validatePlannedSteps(step.body, depth + 1);
        return { id, kind: 'forEach', itemsVar: step.itemsVar, itemVar: step.itemVar, maxItems, body };
      }
      case 'goal': {
        if (typeof step.goal !== 'string' || !step.goal) throw err('PLANNER_INVALID_OUTPUT', `steps[${idx}].goal 非法`);
        const out: FlowStep = { id, kind: 'goal', goal: step.goal, expect: validateExpects(step.expect, `steps[${idx}].expect`) };
        if (typeof step.valuesRef === 'string') (out as { valuesRef?: string }).valuesRef = step.valuesRef;
        return out;
      }
      default:
        throw err('PLANNER_INVALID_OUTPUT', `steps[${idx}].kind 非法: ${String(step.kind)}`);
    }
  });
}

interface ChatMessage {
  role: 'system' | 'user' | 'assistant';
  content: string;
}

/** openai-compatible Provider：/chat/completions + JSON 输出 + 一次受控修复请求（DESIGN §7）。 */
export class OpenAICompatibleProvider implements PlannerProvider {
  constructor(
    private readonly opts: {
      cfg: PlannerConfig;
      /** key 由调用方从 apiKeyEnv 读取后注入；保留在内存，不写日志。 */
      apiKey?: string;
      fetchImpl?: typeof fetch;
    },
  ) {}

  private async chat(messages: ChatMessage[], jsonMode: boolean): Promise<string> {
    const { cfg, apiKey, fetchImpl = fetch } = this.opts;
    if (!apiKey) throw err('PLANNER_NOT_CONFIGURED', `缺少规划模型 key（${cfg.apiKeyEnv}）`);
    const body: Record<string, unknown> = { model: cfg.model, messages, temperature: 0 };
    if (jsonMode) body.response_format = { type: 'json_object' };
    const res = await fetchImpl(`${cfg.baseUrl!.replace(/\/+$/, '')}/chat/completions`, {
      method: 'POST',
      headers: { 'content-type': 'application/json', authorization: `Bearer ${apiKey}` },
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(cfg.timeoutMs),
    });
    if (!res.ok) {
      const text = await res.text().catch(() => '');
      // 某些兼容端点不支持 response_format：退化为纯提示重试一次
      if (res.status === 400 && jsonMode) return this.chat(messages, false);
      throw err('PLANNER_INVALID_OUTPUT', `规划服务 ${res.status}: ${text.slice(0, 200)}`);
    }
    const data = (await res.json()) as { choices?: Array<{ message?: { content?: string } }> };
    return data.choices?.[0]?.message?.content ?? '';
  }

  async plan(input: PlannerInput): Promise<FlowStep[]> {
    const messages: ChatMessage[] = [
      { role: 'system', content: systemPrompt() },
      { role: 'user', content: userPrompt(input) },
    ];
    let content = await this.chat(messages, true);
    let steps: FlowStep[];
    try {
      const parsed = JSON.parse(extractJson(content)) as { steps?: unknown };
      steps = validatePlannedSteps(parsed.steps);
    } catch (first) {
      messages.push({ role: 'assistant', content: content.slice(0, 4000) });
      messages.push({
        role: 'user',
        content: `你的输出未通过校验：${(first as Error).message}。请重新输出符合规则的 JSON {"steps":[...]}。`,
      });
      content = await this.chat(messages, true);
      try {
        const parsed = JSON.parse(extractJson(content)) as { steps?: unknown };
        steps = validatePlannedSteps(parsed.steps);
      } catch (second) {
        throw err('PLANNER_INVALID_OUTPUT', `规划输出两次校验失败: ${(second as Error).message}`);
      }
    }
    return steps;
  }
}

function extractJson(content: string): string {
  const fenced = content.match(/```(?:json)?\s*([\s\S]*?)```/);
  const candidate = fenced ? fenced[1] : content;
  const start = candidate.indexOf('{');
  const end = candidate.lastIndexOf('}');
  if (start < 0 || end <= start) throw err('PLANNER_INVALID_OUTPUT', '输出中找不到 JSON 对象');
  return candidate.slice(start, end + 1);
}

File: tools/jev-desktop/core/planner.py
"""规划器（DESIGN §7）：OpenAI 兼容 /chat/completions，run 模式每轮出 1-3 步严格 JSON。

- 输出经 steps.validate_plan_round 严格校验；校验失败最多一次修复请求；
- 不放宽预算；未配置时 run 返回 PLANNER_NOT_CONFIGURED（不猜模型/供应商）；
- 规划器只见紧凑文本观察（vlm 档才见图），控 token。
"""

import json
import re

import httpx

from .errors import err
from .steps import validate_plan_round

MAX_ROUNDS_STEPS = 3


def planner_configured(cfg: dict, env: dict | None = None) -> bool:
    import os
    env = os.environ if env is None else env
    p = cfg["planner"]
    return bool(p.get("baseUrl") and p.get("model") and env.get(p.get("apiKeyEnv") or ""))


def planner_unconfigured_error(cfg: dict):
    p = cfg["planner"]
    return err("PLANNER_NOT_CONFIGURED",
               "run 模式需要规划 LLM：请在配置中设置 planner.baseUrl / planner.model，"
               f"并在环境变量 {p.get('apiKeyEnv')} 提供密钥（execute 模式不依赖规划器）",
               details={"example": {"planner": {"baseUrl": "https://api.deepseek.com/v1",
                                                "model": "deepseek-chat",
                                                "apiKeyEnv": p.get("apiKeyEnv")}}})


SYSTEM_PROMPT = """你是 Windows 桌面自动化任务规划器。根据目标、当前屏幕快照与最近动作，输出严格 JSON（不要多余文字）：
{"analysis":"一句话分析当前状态","done":布尔,"steps":[1-3个步骤]}
- done=true 表示目标已完成（必须非常有把握；steps 为空数组）
- 步骤 kind 只能是 action|wait|goal|screenshot|extract
- action.act ∈ click|double_click|right_click|type|set_value|press|scroll|select|toggle|expand|collapse|focus|invoke|launch|close
- 定位 target 优先用快照中的 ref：{"kind":"ref","ref":"@快照id:eN"}；其次
  {"kind":"uia","controlType":"Button","name":"保存"}；纯语义目标用 {"kind":"text","target":"描述"}
- type/set_value/press/launch 的步骤必须带 value；scroll 的 value 是 down|up|left|right
- 界面被弹窗/对话框阻挡时，先规划处理弹窗的步骤
- 目标完全无法推进时：done=false 且 steps 只给一个 kind=wait 的步骤
规则：不要发明不存在的 ref；不要输出可执行代码；一次只规划当前界面上的下一步（最多 3 步）。"""


class Planner:
    def __init__(self, cfg: dict, metrics, env: dict | None = None):
        import os
        self.cfg = cfg["planner"]
        self.metrics = metrics
        env = os.environ if env is None else env
        key_env = self.cfg.get("apiKeyEnv") or ""
        self.api_key: str | None = env.get(key_env) or None
        self.count = 0

    def available(self) -> bool:
        return bool(self.api_key and self.cfg.get("baseUrl") and self.cfg.get("model"))

    def _require(self) -> str:
        if not self.available():
            raise err("PLANNER_NOT_CONFIGURED",
                      f"规划器未就绪：需要 planner.baseUrl/planner.model 与环境变量 {self.cfg.get('apiKeyEnv')} 的密钥")
        return self.api_key

    def _chat(self, messages: list[dict], *, json_mode: bool, timeout_s: float) -> str:
        key = self._require()
        self.count += 1
        self.metrics.planner_requests += 1
        body: dict = {"model": self.cfg["model"], "messages": messages,
                      "temperature": self.cfg.get("temperature", 0.2),
                      "max_tokens": self.cfg.get("maxTokens", 2048)}
        if json_mode:
            body["response_format"] = {"type": "json_object"}
        url = self.cfg["baseUrl"].rstrip("/") + "/chat/completions"
        try:
            resp = httpx.post(url, json=body, timeout=timeout_s,
                              headers={"authorization": f"Bearer {key}", "content-type": "application/json"})
        except httpx.HTTPError as e:
            raise err("PROVIDER_ERROR", f"规划服务请求失败: {type(e).__name__}: {e}", retryable=True)
        if resp.status_code == 400 and json_mode:
            return self._chat(messages, json_mode=False, timeout_s=timeout_s)  # 端点不支持 json mode
        if resp.status_code != 200:
            raise err("PROVIDER_ERROR", f"规划服务 {resp.status_code}: {resp.text[:200]}",
                      retryable=resp.status_code in (429, 500, 502, 503, 504))
        data = resp.json()
        usage = data.get("usage") or {}
        # OpenAI 风格 usage：prompt_tokens/completion_tokens
        norm = {"input_tokens": usage.get("prompt_tokens"), "output_tokens": usage.get("completion_tokens")}
        self.metrics.add_tokens(norm)
        try:
            return data["choices"][0]["message"]["content"] or ""
        except Exception as e:
            raise err("PROVIDER_ERROR", f"规划响应解析失败: {e}")

    def plan_round(self, *, goal: str, criteria: list[str], snapshot_text: str,
                   window: dict, recent: list[dict], budget_note: str,
                   timeout_s: float = 60.0) -> dict:
        user = {
            "goal": goal,
            "successCriteria": criteria or ["完成目标的直接可观察结果（未显式配置验收）"],
            "window": window,
            "snapshot": snapshot_text[:6000],
            "recentRounds": recent[-3:],
            "budget": budget_note,
        }
        messages = [
            {"role": "system", "content": SYSTEM_PROMPT},
            {"role": "user", "content": json.dumps(user, ensure_ascii=False)},
        ]
        content = self._chat(messages, json_mode=True, timeout_s=timeout_s)
        try:
            parsed = _extract_json(content)
            return validate_plan_round(parsed, max_steps=MAX_ROUNDS_STEPS)
        except Exception as first:
            messages.append({"role": "assistant", "content": content[:4000]})
            messages.append({"role": "user", "content":
                             f"你的输出未通过校验：{first}。请重新只输出符合规则的 JSON "
                             '{"analysis":"...","done":false,"steps":[...]}。'})
            content = self._chat(messages, json_mode=True, timeout_s=timeout_s)
            try:
                parsed = _extract_json(content)
                return validate_plan_round(parsed, max_steps=MAX_ROUNDS_STEPS)
            except Exception as second:
                raise err("PROVIDER_ERROR", f"规划输出两次校验失败: {second}",
                          details={"firstError": str(first)})


def _extract_json(content: str) -> dict:
    fenced = re.search(r"```(?:json)?\s*([\s\S]*?)```", content)
    candidate = fenced.group(1) if fenced else content
    start, end = candidate.find("{"), candidate.rfind("}")
    if start < 0 or end <= start:
        raise ValueError("输出中找不到 JSON 对象")
    return json.loads(candidate[start:end + 1])
